Support finding EWS runs from GitHub PRs in update-test-expectations

gsnedders opened the pull request:
#### 1a61da068f12ccd67a405590d8022ac115fdd67c
<pre>
Support finding EWS runs from GitHub PRs in update-test-expectations
<a href="https://bugs.webkit.org/show_bug.cgi?id=257158">https://bugs.webkit.org/show_bug.cgi?id=257158</a>

Reviewed by Jonathan Bedard.

This adds support for fetching layout test results based on GitHub PR
statuses. To do this, we refactor test_expectation_updater to be able to
support multiple sources of layout tests results, and add various
subcommands to it. With the subcommands added, we rename the script to
update-test-expectations, and add a legacy
update-test-expectations-from-bugzilla which calls the bugzilla
subcommand.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub._commit_response):
* Tools/Scripts/update-test-expectations: Copied from Tools/Scripts/update-test-expectations-from-bugzilla.
* Tools/Scripts/update-test-expectations-from-bugzilla:
* Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher.py:
(lookup_ews_results_from_pr):
(lookup_ews_results_from_repo_via_pr):
* Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher_unittest.py:
(ResultsFetcherTest.test_is_relevant_results):
(ResultsFetcherTest.test_lookup_ews_results_from_bugzilla):
(ResultsFetcherGitHubTest):
(ResultsFetcherGitHubTest.webserver):
* Tools/Scripts/webkitpy/common/net/bugzilla/test_expectation_updater.py:
(configure_logging.LogHandler):
(argument_parser):
(TestExpectationUpdater.__init__):
(TestExpectationUpdater.fetch_from_bugzilla):
(TestExpectationUpdater):
(TestExpectationUpdater.fetch_from_github_pr):
(TestExpectationUpdater.do_update):
(main):
* Tools/Scripts/webkitpy/common/net/bugzilla/test_expectation_updater_unittest.py:
(TestExpectationUpdaterTest.test_update_test_expectations):

Canonical link: <a href="https://commits.webkit.org/272152@main">https://commits.webkit.org/272152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/454825ba631080141920237183aa7120ed2b2baa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/30780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/9458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/32447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/33275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/27811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/31524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/11765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/6702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/33275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/31093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/11765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/32447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/33275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/30934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/11765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/32447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/34612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/11765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/32447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/34612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/6998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/6702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/34612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/28/builds/30571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/8715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/32447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/7715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3988 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/7554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/pull_request.py:
```diff
@@ -60,6 +60,23 @@ def __repr__(self):
+        def __init__(self, url, status, name=None, description=None):
+            self.url = url
+
+            if status not in ("error", "failure", "pending", "success"):
```
This follows from what GitHub does, and seems to map nicely to what BitBucket does (["Valid values: STOPPED, FAILED, INPROGRESS, SUCCESSFUL"](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-pull-request-id-statuses-get)). That said, I'm a bit confused on the BitBucket side—all our code seems to use `pull-requests` in the URL, whereas all the BitBucket documentation uses `pullrequests`. Am I looking at the wrong documentation?

gsnedders replied:
Oh… Oh. Bitbucket Server and Bitbucket have… different REST APIs.

gsnedders replied:
https://developer.atlassian.com/server/bitbucket/rest/v811/api-group-builds-and-deployments/#api-api-latest-projects-projectkey-repos-repositoryslug-commits-commitid-builds-get I think is what we want.

gsnedders commented on Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher.py:
```diff
@@ -207,3 +208,28 @@ def lookup_ews_results_from_bugzilla(
     return lookup_ews_results_from_bugzilla_patch(patch, bot_filter_name)
+
+
+def lookup_ews_results_from_pr(pr):
+    urls = [status.url for status in pr.statuses]
+    return lookup_ews_results(urls)
```
Given we might have other status checks, do we want to somehow filter this down to things from EWS? Hardcoding the URL seems bad, but it's unclear what else we can do here.

gsnedders commented on Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher.py:
```diff
@@ -207,3 +208,28 @@ def lookup_ews_results_from_bugzilla(
+    return lookup_ews_results(urls)
+
+
+def lookup_ews_results_from_repo(repo_path):
```
The mocks just assume we have `repository` as an argument, right? So maybe we should do that, and construct a `repository` object from the CLI `main`?

JonWBedard commented on Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher.py:
```diff
@@ -207,3 +207,28 @@ def lookup_ews_results_from_bugzilla(
+    source_remote = repository.default_remote
+    remote_repo = repository.remote(name=source_remote)
+    if not remote_repo.pull_requests:
+        msg = "Remote repo doesn't support pull requests"
```
Why not just `raise ValueError` directly?

gsnedders replied:
see https://docs.astral.sh/ruff/rules/raw-string-in-exception/; essentially it avoids the string being duplicated in the traceback, which makes it somewhat harder to read

JonWBedard commented on Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher.py:
```diff
@@ -207,3 +207,28 @@ def lookup_ews_results_from_bugzilla(
+
+    if existing_pr is None:
+        msg = "No PR found for current branch"
+        raise ValueError(msg)
```
Ditto on raising `ValueError(<str>)` directly.

gsnedders committed:
Support finding EWS runs from GitHub PRs in update-test-expectations
https://bugs.webkit.org/show_bug.cgi?id=257158

Reviewed by Jonathan Bedard.

This adds support for fetching layout test results based on GitHub PR
statuses. To do this, we refactor test_expectation_updater to be able to
support multiple sources of layout tests results, and add various
subcommands to it. With the subcommands added, we rename the script to
update-test-expectations, and add a legacy
update-test-expectations-from-bugzilla which calls the bugzilla
subcommand.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub._commit_response):
* Tools/Scripts/update-test-expectations: Copied from Tools/Scripts/update-test-expectations-from-bugzilla.
* Tools/Scripts/update-test-expectations-from-bugzilla:
* Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher.py:
(lookup_ews_results_from_pr):
(lookup_ews_results_from_repo_via_pr):
* Tools/Scripts/webkitpy/common/net/bugzilla/results_fetcher_unittest.py:
(ResultsFetcherTest.test_is_relevant_results):
(ResultsFetcherTest.test_lookup_ews_results_from_bugzilla):
(ResultsFetcherGitHubTest):
(ResultsFetcherGitHubTest.webserver):
* Tools/Scripts/webkitpy/common/net/bugzilla/test_expectation_updater.py:
(configure_logging.LogHandler):
(argument_parser):
(TestExpectationUpdater.__init__):
(TestExpectationUpdater.fetch_from_bugzilla):
(TestExpectationUpdater):
(TestExpectationUpdater.fetch_from_github_pr):
(TestExpectationUpdater.do_update):
(main):
* Tools/Scripts/webkitpy/common/net/bugzilla/test_expectation_updater_unittest.py:
(TestExpectationUpdaterTest.test_update_test_expectations):

Canonical link: https://commits.webkit.org/272152@main
